# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import numpy as np
import pandas as pd

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [2]:
# Run this code:

pokemon = pd.read_csv('../pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [3]:
# Your code here:
pokemon.head()


#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [4]:
# Your code here:
pokemon['Legendary'].value_counts()


False    735
True      65
Name: Legendary, dtype: int64

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [6]:
# Your code here:
print("The mean of total points for legenday Pokemon is:", pokemon[pokemon['Legendary']==True]['Total'].mean())
print("The standard deviation of total points for legenday Pokemon is:", pokemon[pokemon['Legendary']==True]['Total'].std())
print("The mean of total points for non-legenday Pokemon is:", pokemon[pokemon['Legendary']==False]['Total'].mean())
print("The standard deviation of total points for non-legenday Pokemon is:", pokemon[pokemon['Legendary']==False]['Total'].std())

The mean of total points for legenday Pokemon is: 637.3846153846154
The standard deviation of total points for legenday Pokemon is: 60.93738905315344
The mean of total points for non-legenday Pokemon is: 417.21360544217686
The standard deviation of total points for non-legenday Pokemon is: 106.76041745713005


The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [7]:
# Your code here:
from scipy.stats import ttest_ind

legendary_total = pokemon[pokemon['Legendary']==True]['Total']
nonlegendary_total = pokemon[pokemon['Legendary']==False]['Total']
stat, p = ttest_ind(legendary_total, nonlegendary_total, equal_var=False)
stat, p

(25.8335743895517, 9.357954335957446e-47)

What do you conclude from this test? Write your conclusions below.

# Your conclusions here:

<span style="color:green">pvalue = 9.3579e-47 <=0,05, por lo que **rechazamos la hipótesis nula** (H0: mu_legend = mu_nonlegen).</span>
    
* H0: las medias de las muestras son iguales.
* H1: las medias de las muestras son diferentes.


How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [8]:
# Your code here:
pokemon['Type 1'].value_counts()


Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Ground       32
Dragon       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64

In [13]:
# Otra forma diferente de listar el conteo de los diferentes tipos:

pokemon.groupby(by='Type 1')['Type 1'].agg('count').sort_values(ascending=False)

Type 1
Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Rock         44
Electric     44
Ground       32
Dragon       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [18]:
# Your code here:
print("The mean of total points for water Pokemon is:", pokemon[pokemon['Type 1']=='Water']['Total'].mean())
print("The standard deviation of total points for water Pokemon is:", pokemon[pokemon['Type 1']=='Water']['Total'].std())
print("The mean of total points for non-water Pokemon is:", pokemon[pokemon['Type 1']!='Water']['Total'].mean())
print("The standard deviation of total points for non-water Pokemon is:", pokemon[pokemon['Type 1']!='Water']['Total'].std())


The mean of total points for water Pokemon is: 430.45535714285717
The standard deviation of total points for water Pokemon is: 113.1882660643146
The mean of total points for non-water Pokemon is: 435.85901162790697
The standard deviation of total points for non-water Pokemon is: 121.0916823020807


Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

In [20]:
# Your code here:

water_total = pokemon[pokemon['Type 1']=='Water']['Total']
nonwater_total = pokemon[pokemon['Type 1']!='Water']['Total']
stat, p = ttest_ind(water_total, nonwater_total, equal_var=True)
stat, p

(-0.4418547448849676, 0.6587140317488793)

Write your conclusion below.

# Your conclusions here:

pvalue = 0.6587 > 0,05, por lo que **aceptamos la hipótesis nula**:

* H0: las medias de las muestras son iguales.

# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [22]:
# Your code here:
from scipy.stats import ttest_rel

stat, p = ttest_rel(pokemon['Attack'], pokemon['Defense'])
stat, p

(4.325566393330478, 1.7140303479358558e-05)

Describe the results of the test in the cell below.

# Your conclusions here:

pvalue = 1.714e-05 <=0,05, por lo que **rechazamos la hipótesis nula**:

* H0: las medias de las muestras (attack y defense) son iguales.
* H1: las medias de las muestras son diferentes.


We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [23]:
# Your code here:

stat, p = ttest_rel(pokemon['Sp. Atk'], pokemon['Sp. Def'])
stat, p

(0.853986188453353, 0.3933685997548122)

Describe the results of the test in the cell below.

# Your conclusions here:

pvalue = 0.3933 > 0,05, por lo que **aceptamos la hipótesis nula**:

* H0: las medias de las muestras (special attack y special defense) son iguales.
* H1: las medias de las muestras son diferentes.


As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [29]:
# Your code here:
from scipy.stats import ttest_1samp

stats, p = ttest_1samp(pokemon['Attack'], pokemon['Defense'])
stats, p

(0      26.143942
 1      13.943944
 2      -3.484624
 3     -38.341761
 4      31.372513
          ...    
 795   -61.870329
 796   -27.013192
 797    16.558229
 798    16.558229
 799   -35.727476
 Name: Defense, Length: 800, dtype: float64,
 array([2.35729214e-109, 1.03035815e-039, 5.19667456e-004, 2.81264285e-183,
        1.94447097e-141, 9.42613092e-063, 3.83188132e-001, 4.85771138e-120,
        3.83188132e-001, 1.60908692e-031, 3.84376495e-001, 9.69550078e-063,
        8.64205513e-168, 2.73106575e-183, 8.27340461e-078, 5.07003297e-104,
        2.12607841e-208, 5.07003297e-104, 2.50208433e-157, 2.50208433e-157,
        2.50208433e-157, 8.27340461e-078, 5.15516227e-004, 3.84376495e-001,
        2.73106575e-183, 3.62950515e-053, 2.12607841e-208, 1.60908692e-031,
        4.18125063e-136, 1.65696145e-017, 2.50208433e-157, 8.27340461e-078,
        2.19545829e-007, 1.09596434e-114, 2.01028674e-093, 4.43518571e-024,
        6.62167363e-012, 2.50208433e-157, 1.08251351e-067, 8.15530565e-00

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [17]:
# Your code here:



Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [18]:
# Your code here:



Based on a 95% confidence, should we reject the null hypothesis?

In [19]:
# Your answer here:

